src/box/gc.c

```diff
@@ -735,6 +809,238 @@ gc_consumer_name(struct gc_consumer *consumer)
+{
+	int rc = 0;
+	struct txn *txn = txn_begin();
+	if (txn == NULL)
+		return -1;
+	/* No need to wait limbo, so let's set TXN_FORCE_ASYNC here. */
+	txn_set_flags(txn, TXN_FORCE_ASYNC);
+
+	struct gc_consumer *consumer;
+	rlist_foreach_entry(consumer, &gc.consumers, in_consumers) {
+		if (consumer->is_persistent &&
+		    !consumer->is_synced &&
+		    gc_consumer_persist(consumer) != 0)
+			goto fail;
+	}
+	rc = txn_commit(txn);
+	goto out;
+fail:
+	rc = -1;
+	txn_abort(txn);
+out:
+	return rc;
+}
+
+static int
+gc_sync_fiber_f(va_list ap)
+{
+	(void)ap;
+	while (!fiber_is_cancelled()) {
+		fiber_check_gc();
+		int rc = gc_consumers_sync_persistent();
+		if (rc != 0) {
+			say_error("Failed to update persistent WAL GC state");
+			diag_log();
+		}
+		fiber_sleep(TIMEOUT_INFINITY);
+	}
+	return 0;
+}
+
+/** GC consumer definition. */
+struct gc_consumer_def {
+	/** Instance UUID. */
+	struct tt_uuid uuid;
+	/** Instance vclock. */
+	struct vclock vclock;
+	/** Description of object owning this consumer. */
+	const char *object;
+};
+
+/** Mapping from tuple.opts to fields of gc_consumer_def. */
+const struct opt_def gc_consumer_def_opts_reg[] = {
+	OPT_DEF("object", OPT_STRPTR, struct gc_consumer_def, object),
+	OPT_END,
+};
+
+/**
+ * Fill gc_consumer_def with opts from the MsgPack map.
+ * Argument map can be NULL - default options are set in this case.
+ */
+static int
+gc_consumer_def_opts_decode(struct gc_consumer_def *def, const char *map,
+			    struct region *region)
+{
+	def->object = "";
+	if (map == NULL)
+		return 0;
+	return opts_decode(def, gc_consumer_def_opts_reg, &map, region);
+}
+
+/** Build gc_consumer definition from a _gc_consumers' tuple. */
+static struct gc_consumer_def *
+gc_consumer_def_new_from_tuple(struct tuple *tuple, struct region *region)
+{
+	struct gc_consumer_def *def =
+		xregion_alloc_object(region, typeof(*def));
+	memset(def, 0, sizeof(*def));
+	if (tuple_field_uuid(tuple, BOX_GC_CONSUMERS_FIELD_UUID, &def->uuid) != 0)
+		return NULL;
+	if (tt_uuid_is_nil(&def->uuid)) {
+		diag_set(ClientError, ER_INVALID_UUID, tt_uuid_str(&def->uuid));
+		return NULL;
+	}
+	const char *mp_vclock = tuple_field_with_type(
+		tuple, BOX_GC_CONSUMERS_FIELD_VCLOCK, MP_MAP);
+	if (mp_vclock == NULL)
+		return NULL;
+	if (mp_decode_vclock_ignore0(&mp_vclock, &def->vclock) != 0) {
+		diag_set(ClientError, ER_INVALID_VCLOCK);
+		return NULL;
+	}
+	const char *opts = NULL;
+	if (tuple_field(tuple, BOX_GC_CONSUMERS_FIELD_OPTS) != NULL) {
+		opts = tuple_field_with_type(tuple, BOX_GC_CONSUMERS_FIELD_OPTS,
+					     MP_MAP);
+	}
+	if (gc_consumer_def_opts_decode(def, opts, region) != 0)
+		return NULL;
+	return def;
+}
+
+int
+gc_load_consumers(void)
+{
+	if (!gc_has_persistent_consumers())
+		return 0;
+	uint32_t region_svp = region_used(&fiber()->gc);
+	struct space *space = space_by_id(BOX_GC_CONSUMERS_ID);
+	if (space == NULL)
+		return -1;
+	struct index *index = space_index(space, 0);
+	if (index == NULL)
+		return -1;
```

Comment:
Nit: The checks are useless because they should have been performed by `gc_has_persistent_consumers`. Please either drop them or remove the call to `gc_has_persistent_consumers`.

Reply:
Replace the checks with assertions.